confmap/expand.go

```diff
@@ -98,10 +98,17 @@ func findURI(input string) string {
 
 // findAndExpandURI attempts to find and expand the first occurrence of an expandable URI in input. If an expandable URI is found it
 // returns the input with the URI expanded, true and nil. Otherwise, it returns the unchanged input, false and the expanding error.
+// This method expects input to start with ${ and end with }
 func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (any, bool, error) {
 	uri := findURI(input)
 	if uri == "" {
-		// No URI found, return.
+		if mr.defaultProvider != nil {
```

Comment:
Could we move this code into `findURI` (which will probably require making it a method on `*Resolver`) and just use the returned `uri` in our regular flow? Adding an exception like this instead of baking the concept of a default scheme into our resolution process makes me worried we are opening ourselves up to missing cases in the future.

In particular, it looks like `findURI` checks whether a scheme is provided, and I think we should have logic that says "use the provided scheme, or a default scheme if one has been defined".

Reply:
I'll look at that option again, but I believe it will mean returning a string by combining the default provider's scheme and the current input.  So if the input is `${PORT}`, `findURI` needs to return `${env:PORT}` and I was trying to avoid messing with the input string too much.

Reply:
> I was trying to avoid messing with the input string too much.

Are there any specific problems you are trying to avoid? IMO just adding the prefix should be fine and would make the code cleaner

Reply:
We're also need to be careful (I think) about using the default provider on nested URIs

Reply:
Also that will cause a problem lower in `findAndExpandURI` when the identified uri is used as the exact replace string from the input

Reply:
Should we return some structured data? Some struct that contains the raw URI and has methods to get the full URI (i.e. adding the default if needed), and the provider?

Reply:
@mx-psi @evan-bradley I've pushed an experiment with findURI that allows expanding stuff like `${env:CUSTOM${NAME}}`.  If we like the approach/extra functionality I will add a lot more unit tests.  For now I've updated the existing to match the new behavior when a DefaultProvider is set.

Reply:
> keep all the defaulting logic contained to a single function that's separated from the actual resolution logic

This ends up being difficult bc of the support for nested URIs and the default provider not being in the `providers` list and not requiring a `location` since we know the uri.  I can keep experimenting, but unless we add the DefaultProvider to the providers map we wont be able to use 

```go
                lURI, locErr := newLocation(uri)
		if locErr != nil {
			return nil, false, locErr
		}
		if strings.Contains(lURI.opaqueValue, "$") {
			return nil, false, fmt.Errorf("the uri %q contains unsupported characters ('$')", lURI.asString())
		}
		ret, err = mr.retrieveValue(ctx, **lURI)**
```

exactly bc `mr.retrieveValue` expects the sheme to be in the providers map

Reply:
I've pushed another change that utilizes DefaultProvider being in the provider map and modifies the uri returned from `findURI` to include the default provider scheme.  This reduces the number of changes needed in other methods in the resolution logic, but feels a little magical to me.  Maybe a struct returned from `findURI` instead of 2 strings would be better?